### PR TITLE
Bazel core sanitization

### PR DESCRIPTION
**Pull request by mhevery**
NOTE: This PR best reviewed one SHA at a time

## refactor(core): add `packages/core/sanitization` bazel target

1. Breaking up monolithic build targets into smaller builds is desirable.
2. This is motivated by code duplication between `packages/core` and `packages/compiler`. By breaking `packages/core` into smaller sub-builds it will be possible for the `packages/compiler` to directly depend on `packages/core/src/sanitization` which will than allow removal of code duplication.

While creating the smaller bundle it was discovered that these two files needed to be moved:
- `packages/core/src/sanitization/sanitizer.ts`: This file was depending on DI and as such it would prevent `packages/core/src/sanitization` from being a leaf target. Moved to `packages/cor/src/sanitizer/sanitizer.ts`
- `packages/core/src/sanitization/sanitization.ts`: This file was depending on render3 directory and as such it would prevent `packages/core/src/sanitization` from being a leaf target. Moved to `packages/cor/src/render3/sanitization.ts`. The contents of this file contains Ivy instructions and so the new location is a better fit for it.

## refactor: Remove `compiler`/`core` code duplication

The `packages/core` depends on `packages/compiler`, therefore `packgase/compiler` can not depend on `packgase/core` or it would create circular dependency.

In previous SHA `packages/core/src/sanitization` sub-build was create to allow code reuse. This allows `packages/compiler` to depend on `packages/core/src/sanitization` and reuse code.

NOTE: This is a proof of concept PR, there are more places where code reuse could be implemented.

---

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/sanitization/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,19 @@
+load("//tools:defaults.bzl", "ts_library")
+
+package(default_visibility = [
+    "//packages/core:__subpackages__",
+    "//tools/public_api_guard:__pkg__",
```
Comment (IgorMinar) [inline packages/core/src/sanitization/BUILD.bazel]: I don't believe that you need this. Can you remove?

File: packages/core/src/sanitization/url_sanitizer.ts [hunk 2]
```diff
@@ -36,9 +36,6 @@ import {isDevMode} from '../util/is_dev_mode';
  */
 const SAFE_URL_PATTERN = /^(?:(?:https?|mailto|ftp|tel|file):|[^&:/?#]*(?:[/?#]|$))/gi;
 
-/* A pattern that matches safe srcset values */
-const SAFE_SRCSET_PATTERN = /^(?:(?:https?|file):|[^&:/?#]*(?:[/?#]|$))/gi;
-
```
Comment (IgorMinar) [inline packages/core/src/sanitization/url_sanitizer.ts]: yay! dead code elimination!

File: packages/compiler/src/render3/view/styling_builder.ts [hunk 3]
```diff
@@ -5,6 +5,8 @@
  * Use of this source code is governed by an MIT-style license that can be
  * found in the LICENSE file at https://angular.io/license
  */
+import {isStyleSanitizable} from '@angular/core/src/sanitization/url_sanitizer';
```
Comment (IgorMinar) [inline packages/compiler/src/render3/view/styling_builder.ts]: this is a deep import into a path that doesn't exist at runtime.

keep in mind that we'd need to support at least 2 different runtimes:
- node.js
- browser

and for both runtimes (now that node.js supports or will soon support ESM) we need to support two module formats:
- UMD
- ESM

Deep imports are simply a no go with the current set of packaging constraints we have, which leaves you with the only known viable option to be private exports, but that doesn't solve the compilation issues breaking things up into smaller compilation units.

**Review by IgorMinar — CHANGES_REQUESTED** [review]
Deep import across entry-points won't work as proposed.

It might still be worth breaking up the build targets though

File: packages/compiler/src/render3/view/styling_builder.ts (hunk 3 above)
Comment (alxhub) [reply packages/compiler/src/render3/view/styling_builder.ts]: We could have an entrypoint `@packages/core/api` or similar, which is published across all formats and only built with `tsc`, that the compiler _could_ depend on, I believe. Would that be a feasible alternative?

The last time I checked though this was challenging to do using our Bazel packaging system.

File: packages/core/src/sanitization/BUILD.bazel (hunk 1 above)
Comment (mhevery) [reply packages/core/src/sanitization/BUILD.bazel]: removed

File: packages/compiler/src/render3/view/styling_builder.ts (hunk 3 above)
Comment (mhevery) [reply packages/compiler/src/render3/view/styling_builder.ts]: What do you suggest as a solution then?

Comment (mhevery) [reply packages/compiler/src/render3/view/styling_builder.ts]: Why does there have to be a package? Is there a way to tell the rollup to just inline this particular import?

Comment (mhevery) [reply packages/compiler/src/render3/view/styling_builder.ts]: What about changing it to `import {isStyleSanitizable} from '../../../../core/src/sanitization/url_sanitizer';` would that tell rollup to just inline ?

**Pull request closed without merging**